src/sources/hydra.rs

```diff
@@ -6,37 +6,66 @@ use bytes::Bytes;
 
 use pallas::network::miniprotocols::Point;
 use pallas::ledger::addresses::Address;
-
-use pallas::interop::utxorpc::spec::cardano::{TxOutput, TxInput, Tx};
+use pallas::interop::utxorpc::spec::cardano::{TxOutput, Tx};
 
 use gasket::framework::*;
-use serde::{Deserialize, Serialize, Deserializer};
-use tracing::{debug, info, warn};
+use tracing::{debug, warn};
 
 use tokio_tungstenite::WebSocketStream;
 use futures_util::StreamExt;
 use tokio_tungstenite::{connect_async, tungstenite::protocol::Message};
 
+use serde::{Deserialize, Serialize, Deserializer};
+use serde_json::Value;
+use serde::de::{self};
+
 use crate::framework::*;
 
+// Models /most/ hydra messages. Messages without head_id are excluded.
+pub struct HydraMessage {
+    seq: u64,
+    head_id: Vec<u8>,
+    payload: HydraMessagePayload
+}
+
+impl<'de> Deserialize<'de> for HydraMessage {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        let map: Value = Deserialize::deserialize(deserializer)?;
+
+        let seq = map.get("seq")
+            .ok_or_else(|| de::Error::missing_field("seq"))?
+            .as_u64()
+            .ok_or_else(|| de::Error::custom("seq should be a u64"))?;
+
+        let head_id_str = map.get("headId")
+            .ok_or_else(|| de::Error::missing_field("headId"))?
+            .as_str()
+            .ok_or_else(|| de::Error::custom("headId should be a string"))?;
+
+        let head_id = hex::decode(head_id_str)
+            .map_err(|_e| serde::de::Error::custom(format!("Expected hex-encoded headId")))?;
+
+        let payload = HydraMessagePayload::deserialize(&map)
+            .map_err(de::Error::custom)?;
+
+        Ok(HydraMessage { seq, head_id, payload })
+    }
+}
 
 #[derive(Serialize, Deserialize, Debug, Clone)]
 #[serde(tag = "tag", rename_all = "PascalCase")]
-pub enum HydraMessage {
+pub enum HydraMessagePayload {
```

Comment:
after this change test probably does not compile. You need to change `HydraMessage::SomethingElse` to `HydraMessagePayload::Other`